plugins/teststeps/fileupload/awsFileUpload.go

```diff
@@ -0,0 +1,279 @@
+	"github.com/facebookincubator/contest/plugins/teststeps"
+	"github.com/prometheus/common/log"
+)
+
+// Define some constants for uploading things into a S3 bucket
+const (
+	S3_REGION = "" // your aws server region
+	S3_BUCKET = "" // your aws bucket
+	S3_PATH   = "" // your path where to upload the file
+)
+
+// Name is the name used to look this plugin up.
+var Name = "awsFileUpload"
+
+// Event names for this plugin.
+const (
+	EventURLtoReport = event.Name("URLtoReport")
+)
+
+// Events defines the events that a TestStep is allow to emit
+var Events = []event.Name{
+	EventURLtoReport,
+}
+
+// Cmd is used to run arbitrary commands as test steps.
+type FileUpload struct {
+	path            *test.Param // Path to file that shall be uploaded
+	filename        *test.Param // Filename to file that shall be uploaded
+	emitURLtoReport bool        // Bool that defines that URLtoReport Event shall be triggered
+	compress        bool        // Bool that defines if the data should be compressed before uploading it
+}
+
+// Datatype for the emmiting the URLtoReport Event
+type eventURLtoReportPayload struct {
+	Msg string
+}
+
+// Name returns the plugin name.
+func (ts FileUpload) Name() string {
+	return Name
+}
+
+func emitEvent(ctx xcontext.Context, name event.Name, payload interface{}, tgt *target.Target, ev testevent.Emitter) error {
+	payloadStr, err := json.Marshal(payload)
+	if err != nil {
+		return fmt.Errorf("cannot encode payload for event '%s': %v", name, err)
+	}
+	rm := json.RawMessage(payloadStr)
+	evData := testevent.Data{
+		EventName: name,
+		Target:    tgt,
+		Payload:   &rm,
+	}
+	if err := ev.Emit(ctx, evData); err != nil {
+		return fmt.Errorf("cannot emit event EventCmdStart: %v", err)
+	}
+	return nil
+}
+
+// Run executes the cmd step.
+func (ts *FileUpload) Run(ctx xcontext.Context, ch test.TestStepChannels, params test.TestStepParameters,
+	ev testevent.Emitter, resumeState json.RawMessage) (json.RawMessage, error) {
+
+	if err := ts.validateAndPopulate(params); err != nil {
+		return nil, err
+	}
+	f := func(ctx xcontext.Context, target *target.Target) error {
+		// expand args
+		path, err := ts.path.Expand(target)
+		if err != nil {
+			return fmt.Errorf("failed to expand argument '%s': %v", ts.path, err)
+		}
+		filename, err := ts.filename.Expand(target)
+		if err != nil {
+			return fmt.Errorf("failed to expand argument dir '%s': %v", ts.filename, err)
+		}
+		// Compress if compress parameter is true
+		if ts.compress {
+			// Create output file
+			newpath := path + ".tar.gz"
+			tar_gz, err := os.Create(newpath)
+			if err != nil {
+				return fmt.Errorf("error writing archive: %w", err)
+			}
+			defer tar_gz.Close()
+			// Create the archive and write the output to the "out" Writer
+			err = createTarArchive(path, tar_gz)
+			if err != nil {
+				log.Fatalln("Error creating archive:", err)
+			}
+			fmt.Println("Tar archive created successfully")
+			path = newpath
+			filename = filename + ".tar.gz"
```

Comment:
Where will it put the compressed file, next to the original uncompressed one?
I guess that is ok. I was just wondering, since later we have to read the entire file into memory for upload anyway, should we just write the archive into a byte buffer and then upload it, so basically without creating a `tar.gz` file that needs to be cleaned up? I think that would be nicer